### PR TITLE
MSE-in-Workers draft feature specification

### PR DESCRIPTION
**Pull request by wolenetz**
Specifies the exposure of MSE API additionally to DedicatedWorkerGlobalScope.
Updates, when MSE is in such a worker scope, cross-context communication between a Window HTMLMediaElement and a DedicatedWorker MSE instance to make it clear to implementors what is necessary for interop on this feature.
Spec feature for MSE-in-Workers is #175 

Note that Chromium already has an experimental implementation and there is a demo of the feature with instructions available at https://wolenetz.github.io/mse-in-workers-demo/mse-in-workers-demo.html

***
<a href="https://pr-preview.s3.amazonaws.com/wolenetz/media-source/pull/282.html" title="Last updated on Jul 26, 2021, 10:35 PM UTC (077e233)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/media-source/282/801c698...wolenetz:077e233.html" title="Last updated on Jul 26, 2021, 10:35 PM UTC (077e233)">Diff</a>

### CODE REVIEW

**Comment by wolenetz** [comment]
It looks like I need to fix a few spec-prod errs (div in dl, dl in ol, div in ol, div in ol, ol in dl) per https://github.com/w3c/media-source/runs/2947876193?check_suite_focus=true
Please don't let those block review.

File: media-source-respec.html [hunk 1]
```diff
@@ -418,73 +422,158 @@ <h3>Attributes</h3>
               </ol>
             </li>
 
-            <li>Let <var>SourceBuffer audioTracks list</var> equal the {{AudioTrackList}} object returned by <var>sourceBuffer</var>.{{SourceBuffer/audioTracks}}.</li>
-            <li>If the <var>SourceBuffer audioTracks list</var> is not empty, then run the following steps:
+            <li>Let |SourceBuffer audioTracks list| equal the {{AudioTrackList}} object returned by
```
Comment (tidoust) [inline media-source-respec.html:425]: It is a good idea to use the `|` shorthand syntax instead of `<var>`. This makes ReSpec track and highlight variables when user hovers or clicks on them as in https://user-images.githubusercontent.com/8426945/74150174-5adf9f00-4c2f-11ea-82ee-d7c059a8f772.gif

That may exceed the scope of this PR but I note that it would be equally good to replace constructs such as `<var>sourceBuffer</var>` with `|sourceBuffer|` throughout algorithms as well.

File: media-source-respec.html [hunk 2]
```diff
@@ -369,6 +369,10 @@ <h3>Attributes</h3>
           <p>The event handler for the {{sourceended}} event.</p>
         </dd><dt><dfn><code>onsourceclose</code></dfn> of type {{EventHandler}}</dt><dd>
           <p>The event handler for the {{sourceclose}} event.</p>
+        </dd>
+        <dt><dfn><code>canConstructInDedicatedWorker</code></dfn> of type {{boolean}}</dt><dd>
+          <p>Returns true.</p>
+          <p class="note">This attribute enables feature detection of support for MSE-in-Workers without requiring an application to attempt usage first from a DedicatedWorker.</p>
```
Comment (tidoust) [inline media-source-respec.html:375]: "MSE-in-Workers" and "DedicatedWorker" seem jargon to me. Suggest to expand with regular prose as in:

```suggestion
          <p class="note">This attribute enables feature detection of support for creating and using a {{MediaSource}} object in a dedicated worker directly from the main thread without requiring an application to attempt usage first from a dedicated worker.</p>
```

File: media-source-respec.html [hunk 3]
```diff
@@ -599,6 +688,40 @@ <h3>Event Summary</h3>
...
+        in the context where it was created.</p>
+
+        The rest of this section describes a model for bounding information latency for attachments of a {{Window}}
+        media element to a {{DedicatedWorkerGlobalScope}} {{MediaSource}}. While the model describes communication using
+        message passing, implementations MAY choose to communicate in potentially faster ways, such as using shared
+        memory and locks. Attachments to a {{Window}} {{MediaSource}} synchronously have the information already without
+        communicating it across contexts.
+
+        <p>A {{MediaSource}} that is constructed in a {{DedicatedWorkerGlobalScope}} has an internal, unexposed
+        <dfn id="port-to-main">port to main</dfn> nullable variable that stores a {{MessagePort}} setup during
```
Comment (tidoust) [inline media-source-respec.html:711]: In other specs, these internal, unexposed variable are called [internal slots](https://medium.com/jspoint/what-are-internal-slots-and-internal-methods-in-javascript-f2f0f6b38de) and rendered as `[[port to main]]`. From a ReSpec perspective, dfn would be `<dfn data-dfn-for="MediaSource">[[\port to main]]</dfn> and references would use `{{MediaSource/[[port to main]]}}`

File: media-source-respec.html [hunk 4]
```diff
@@ -599,6 +688,40 @@ <h3>Event Summary</h3>
...
+        attachment and nulled during detachment. A {{Window}} {{MediaSource}}'s [=port to main=] is always null.</p>
+
+        </p>An {{HTMLMediaElement}} extended by this specification and attached to a {{DedicatedWorkerGlobalScope}}
+        {{MediaSource}} similarly has an internal, unexposed <dfn id="port-to-worker">port to worker</dfn> nullable
+        variable that stores a {{MessagePort}} setup during attachment and nulled during detachment. A
+        [=port to worker=] is always null unless attached to a {{DedicatedWorkerGlobalScope}} {{MediaSource}}.</p>
+
```
Comment (tidoust) [inline media-source-respec.html:718]: I would also add the definition of the `channel` internal slot here as well, even though it is only going to be set in objects created in dedicated workers.

File: media-source-respec.html [hunk 5]
```diff
@@ -755,7 +918,16 @@ <h4><dfn>SourceBuffer Monitoring</dfn></h4>
 
         <section id="active-source-buffer-changes">
           <h4><dfn>Changes to selected/enabled track state</dfn></h4>
-          <p>During playback {{MediaSource/activeSourceBuffers}} needs to be updated if the <a def-id="videoref" name="dom-videotrack-selected">selected video track</a>, the <a def-id="videoref" name="dom-audiotrack-enabled">enabled audio track(s)</a>, or a text track <a def-id="videoref" name="dom-texttrack-mode">mode</a> changes. When one or more of these changes occur the following steps need to be followed.</p>
+          <p>During playback {{MediaSource/activeSourceBuffers}} needs to be updated if the <a def-id="videoref" name="dom-videotrack-selected">selected video track</a>, the <a def-id="videoref" name="dom-audiotrack-enabled">enabled audio track(s)</a>, or a text track <a def-id="videoref" name="dom-texttrack-mode">mode</a> changes. When one or more of these changes occur the following steps need to be followed.
+             Also, when {{MediaSource}} was constructed in {{DedicatedWorkerGlobalScope}}, then each change that occurs
```
Comment (tidoust) [inline media-source-respec.html:922]:
```suggestion
             Also, when {{MediaSource}} was constructed in a {{DedicatedWorkerGlobalScope}}, then each change that occurs
```

**Review by tidoust — COMMENTED** [review]
Some comments on the syntax.

The logic looks sound on the surface (a bit convoluted but it's not obvious that it can be simplified...). I haven't tried to look into details yet.

**Review by tidoust — COMMENTED** [review]
Looking into it some more, I wonder what are the benefits of specifying such a detailed message passing logic between the object in the worker and its mirror in Window.

Wouldn't it be enough to mention the need for implementation to mirror information between Window and worker objects using whatever implementation means they want (as alluded to in Cross-context communication model section) and keep algorithms somewhat unchanged except perhaps to note which information needs to have been mirrored and/or where the algorithm step runs?

For instance, considering the following steps:

> If the MediaSource was constructed in a DedicatedWorkerGlobalScope : Post an internal end of stream message to port to main whose implicit handler in Window notifies the media element that it now has all of the media data. 
>
> Otherwise: Notify the media element that it now has all of the media data. 

Why not keep it to what it was before the change?

> Notify the media element that it now has all of the media data.

There is only one media element. It is clear that it is attached to the Window context. The term "notify" is vague, allowing arbitrary complex implementations. I do realize that the worker case is going to be more complex to implement but, from a spec perspective, I'm not clear why specifying the extra step that cannot be directly tested in practice is needed.

(There are likely other algorithms where it remains useful to clarify what is to happen, for instance to mention that visible VideoTrack objects attached to a media element are going to be mirrors of the ones in the worker)

File: media-source-respec.html (hunk 1 above)
Comment (wolenetz) [reply media-source-respec.html:425]: I agree - I used `|` in much of the PR. I'll fix portions in this PR that touch lines with `<var>`, too. There are currently over 300 uses, so fixing the remaining bulk will be a different PR.

Comment (wolenetz) [reply media-source-respec.html]: cf47a8f addresses this feedback. There still remain over 200 occurrences in the spec after that.

File: media-source-respec.html (hunk 2 above)
Comment (wolenetz) [reply media-source-respec.html]: Good point. 9efdf52 addresses this comment.

**Comment by wolenetz** [comment]
> Looking into it some more, I wonder what are the benefits of specifying such a detailed message passing logic between the object in the worker and its mirror in Window.
> 
> Wouldn't it be enough to mention the need for implementation to mirror information between Window and worker objects using whatever implementation means they want (as alluded to in Cross-context communication model section) and keep algorithms somewhat unchanged except perhaps to note which information needs to have been mirrored and/or where the algorithm step runs?

The notion of "mirrored" can be interpreted differently in concurrent systems like the main/worker contexts in this feature. For clarity of what information coherency/latency flexibility is allowed, I included the "Cross-context communication model" and use of it (`port to worker` / `port to main`) in specific portions of the algorithms where the communication.

> For instance, considering the following steps:
> 
> > If the MediaSource was constructed in a DedicatedWorkerGlobalScope : Post an internal end of stream message to port to main whose implicit handler in Window notifies the media element that it now has all of the media data.
> > Otherwise: Notify the media element that it now has all of the media data.
> 
> Why not keep it to what it was before the change?
> 
> > Notify the media element that it now has all of the media data.
> 
> There is only one media element. It is clear that it is attached to the Window context. The term "notify" is vague, allowing arbitrary complex implementations. I do realize that the worker case is going to be more complex to implement but, from a spec perspective, I'm not clear why specifying the extra step that cannot be directly tested in practice is needed.

This example is a good one for your case. The "notify" is indeed vague due to legacy media pipelines and demuxers introducing some information latencies. I will revert the convoluted text specifically for this example. In contrast, the buffered and seekable extensions need more clarity because they are synchronous (on main thread) and reflect current information immediately, so clarify what the information latency limits here with the model and its usage seemed to me to be not only a good idea but required for better interoperability.

> (There are likely other algorithms where it remains useful to clarify what is to happen, for instance to mention that visible VideoTrack objects attached to a media element are going to be mirrors of the ones in the worker)

Indeed, like the seekable and buffered extensions, the initialization segment received algorithm and removeSourceBuffer's management of tracks is IMHO made much clearer for interop by the detailed spec. Implementors are free to replace internal, unexposed items in the spec like 'update track state' messages with equivalent.

tl;dr: I wanted to clarify the expectations for better implementation interop.

File: media-source-respec.html (hunk 3 above)
Comment (wolenetz) [reply media-source-respec.html:723]: Yes I've seen the internal slot spec abstraction occurring recently in other specs. I'll update this feature draft for port-to-main/window and channel to be held in explicit internal slots. Separately, in a different PR, I'll update the existing object variables like "first initialization segment received flag" to be held in internal slots, too.

File: media-source-respec.html (hunk 4 above)
Comment (wolenetz) [reply media-source-respec.html:730]: I agree and will update.

File: media-source-respec.html (hunk 5 above)
Comment (wolenetz) [reply media-source-respec.html]: Thanks! Fixed by 82070b2

**Comment by tidoust** [comment]
> tl;dr: I wanted to clarify the expectations for better implementation interop.

Right and you shouldn't consider my comments as blocking in any way! I didn't spot anything wrong in the pull request.

The algorithms can be iterated over to drop bits that won't be observable in any way. These cases will typically appear when test cases get added to the test suite to assess interop between implementations of MSE in workers.

File: media-source-respec.html (hunk 3 above)
Comment (wolenetz) [reply media-source-respec.html]: 98fe6fc should fix this.

File: media-source-respec.html (hunk 4 above)
Comment (wolenetz) [reply media-source-respec.html:732]: 98fe6fc should fix this.

**Comment by wolenetz** [comment]
> The algorithms can be iterated over to drop bits that won't be observable in any way. These cases will typically appear when test cases get added to the test suite to assess interop between implementations of MSE in workers.

Sounds good to me. I've added ee8fc38 to address the one known place where there was perhaps too much verbosity added.

**Comment by wolenetz** [comment]
077e233 is a squashed replacement of the previous chain of commits, to ease potential rebasing.

**Comment by wolenetz** [comment]
@mwatson2 please take a look. Known related work that I'd like to get landed soon, separately, is adding privacy and security sections, and especially (in response to a high-res timer attack concern raised in a response to my intent-to-experiment request in chromium/blink) clarify that the implementation's choice of cross-context communication mechanism needs to be resilient against such attacks (such as providing variable delay akin to postMessage, as the model already describes, if optimizations are made) for information such as MediaSource duration, live-seekable-range (as reported to HTMLMediaElement from worker) and recent error status (as reported to worker MSE from window HTMLMediaElement).

@tidoust please help verify that the changes in response to your comments are correct.

Thank you,
Matt

**Comment by tidoust** [comment]
> @tidoust please help verify that the changes in response to your comments are correct.

That looks good to me, thanks!

I note that the IPR bot complains because Google (and other companies) needs to re-join the working group following re-chartering.

**Review by mwatson2 — APPROVED** [review]

**Comment by mwatson2** [comment]
@wolenetz Looks good to me. My only comments are cosmetic suggestions - not technical - and could be done later, if done at all.

**Comment by wolenetz** [comment]
@ https://github.com/w3c/media-source/pull/282#issuecomment-89805456 (@mwatson2) - Thank you for your review. In general, I agree with your comments. The "potentially faster" text was included to ensure that *at least* no more than the latency of a postMessage() done by the app would be the limit. Otherwise, apps would need to take yet more care to understand the state of the attachment cross-context. In this regard, I'm also currently undoing some of the optimizations taken in the Chromium prototype of this feature, as they may enable Spectre-like exploits.

Since my AC rep hasn't yet rejoined the WG, I'll await until after my vacation (all of next week) to address comments (or land as-is and do a follow-up to address comments).

**Comment by wolenetz** [comment]
@https://github.com/w3c/media-source/pull/282#issuecomment-898054566 in the interests of more easily unblocking the follow-on commits that I'd been working on while pending rejoin to the WG, I'll land this as is and follow-up in later PR on your comments, above.

File: media-source-respec.html [hunk 6]
```diff
@@ -404,108 +411,202 @@ <h3>Attributes</h3>
...
                     </li>
-                    <li>Remove the {{AudioTrack}} object from the <var>SourceBuffer audioTracks list</var>.
-                    <p class="note">
-                      This should trigger {{AudioTrackList}} [[HTML]] logic to <a def-id="queue-and-fire-media-removetrack-with-track-attr-initialized-to"></a> the {{AudioTrack}} object, at the <var>SourceBuffer audioTracks list</var>.
-                      If the {{AudioTrack/enabled}} attribute on the {{AudioTrack}} object was true at the beginning of this removal step, then this should also trigger {{AudioTrackList}} [[HTML]] logic to [=queue a task=] to [=fire an event=] named <a def-id="mediatracklist-change"></a> at the <var>SourceBuffer audioTracks list</var>
-                    </p>
+
+                    <li>
+                      <dl class="switch">
+                        <dt>If the {{MediaSource}} was constructed in a {{DedicatedWorkerGlobalScope}}:</dt>
```
[the remark this answers is not in the document]
Comment (wolenetz) [reply media-source-respec.html:451]: > IIUC, the purpose of the mirror messages is to ensure that the HTML Media Element update steps run as their own task on the main event loop, rather than the associated changes happening "magically" in the middle of some other task execution, right ?

This is typically the case, yes. Especially will be more so once I update this to likely limit any optimizations of the 'cross-context communication mechanism' for web platform security and interop.

[the remark this answers is not in the document]
Comment (wolenetz) [reply media-source-respec.html:451]: > This is by no means an essential change, but I think this would be clearer if the "Otherwise" steps below were broken out an given a name (e.g. "_Remove an audio track from Media Element audio track list_") and then this step can be something like
> 
> > If the MediaSource was constructed in a DedicatedWorkerGlobalScope, post an internal remove track mirror message ... whose implicit handler runs the _Remove an audio track from Media Element audio track list_ algorithm on ....
> > Otherwise run the _Remove an audio track from Media Element audio track list_ algorithm on ....

I'm looking to see if there's a way to keep such a "named step" within the method algorithm steps instead of being moved to an explicit algorithm in the SourceBuffer algorithm's section. Do you have any preference on this?

[the remark this answers is not in the document]
Comment (wolenetz) [reply media-source-respec.html:451]: For consistency with other media WG specs, and internal consistency with existing MSE spec, moving such a "named step" to be an explicit algorithm is probably best.

[the remark this answers is not in the document]
Comment (wolenetz) [reply media-source-respec.html:451]: Mirror if necessary seems the better approach for this, as it is more general.
